### PR TITLE
CRM457-1487 - Add analytics user, views and permissions

### PR DESCRIPTION
**Pull request by dwhenry**
## Description of change
Initial thoughts on analytics user and tables creation

## Link to relevant ticket

[link to ticket](https://dsdmoj.atlassian.net/browse/CRM457-1487)

```sql
select * from all_events limit 3

                  id                  | application_type |                                                                                                                                                                                                                            event
--------------------------------------+------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | {"id": "674ad9b4-d550-48a2-b90e-ea4562a37ef9", "public": false, "details": {}, "linked_id": null, "created_at": "2024-05-31T10:41:13.683Z", "event_type": "new_version", "updated_at": "2024-05-31T10:41:13.683Z", "linked_type": null, "primary_user_id": null, "secondary_user_id": null, "submission_version": 1}
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | {"id": "9fb23351-b1e1-4389-b5e3-a066437d3a41", "public": false, "details": {"comment": "aaa"}, "linked_id": null, "created_at": "2024-05-31T10:41:46.110Z", "event_type": "assignment", "updated_at": "2024-05-31T10:41:46.110Z", "linked_type": null, "primary_user_id": "c4981db1-58c8-458b-8ae4-ba925f3bc8df", "secondary_user_id": null, "submission_version": 1}
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | {"id": "0217b716-e29e-41e5-b146-3560639fb27c", "public": false, "details": {"to": 70, "from": 61, "field": "period", "change": 9, "comment": "aaaa"}, "linked_id": "3eda3f3e-7966-435d-91ea-36254c91b9a8", "created_at": "2024-05-31T10:41:59.765Z", "event_type": "edit", "updated_at": "2024-05-31T10:41:59.765Z", "linked_type": "quotes", "primary_user_id": "c4981db1-58c8-458b-8ae4-ba925f3bc8df", "secondary_user_id": null, "submission_version": 1}
```

```sql
select * from version_events

                  id                  | application_type |   status   |        event_at
--------------------------------------+------------------+------------+-------------------------
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | submitted  | 2024-05-31 10:41:13.683
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | part_grant | 2024-05-31 10:43:16.482
 d85dd60d-b716-4b6d-ad35-2a602e9c143b | crm4             | submitted  | 2024-06-03 09:54:42.996
```

> NOTE: This last one is a WIP as we don't currently store the creation time for the record from Provider so can't calculate how long it took to submit.
```sql
select * from version_events_with_time;
                  id                  | application_type |   status   |        event_at         |  event_time
--------------------------------------+------------------+------------+-------------------------+--------------
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | submitted  | 2024-05-31 10:41:13.683 |
 c96ed6dd-cae9-4855-8681-696e6bd70ccb | crm4             | part_grant | 2024-05-31 10:43:16.482 | 00:02:02.799
 d85dd60d-b716-4b6d-ad35-2a602e9c143b | crm4             | submitted  | 2024-06-03 09:54:42.996 |
```

### CODE REVIEW

**Review by imtiazAhmd — APPROVED** [review]

File: db/migrate/20240611091030_create_all_events.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class CreateAllEvents < ActiveRecord::Migration[7.1]
+  def change
+    create_view :all_events
+  end
+end
```
Comment (imtiazAhmd) [inline db/migrate/20240611091030_create_all_events.rb:5]: i thought that scenic would allow us to keep a different dir/file structure for migrations? Joel, on behalf of imtiaz via tuple

Comment (dwhenry) [reply db/migrate/20240611091030_create_all_events.rb:5]: so you get a different directory for the SQL definitions, but it still needs to create a migration file to apply, this allow them to leverage the existing versioning stuff... not 100% sure I agree, but also not terrible..

File: .env.development [hunk 2]
```diff
@@ -17,3 +17,11 @@ TENANT_ID=not-a-real-secret
...
 SIDEKIQ_WEB_UI_USERNAME=not-a-real-secret
 SIDEKIQ_WEB_UI_PASSWORD=not-a-real-secret
+
+ANALYTICS_USERNAME=analytics_user
+ANALYTICS_PASSWORD=analytics_user_password
+
+kubectl create secret generic rds-postgresql-instance-output \
+    --from-literal=analytics_user=analytics_user \
+    --from-literal=analytics_password='QjAnGA3spw5b3o2' \
+    --namespace=laa-submit-crime-forms-dev
```
Comment (imtiazAhmd) [inline .env.development:27]: is this here by accident?

File: lib/analytics_creator.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  def create_user
+    raise "ANALYTICS_DB_PASSWORD must be set." if password.blank?
+
+    roles = execute("select * from pg_catalog.pg_roles where rolname='#{username}'")
+    return if roles.rows.any?
+
+    execute("CREATE ROLE #{username} WITH LOGIN PASSWORD '#{password}'")
+    execute("GRANT CONNECT ON DATABASE #{database_name} TO #{username}")
+    execute("GRANT USAGE ON SCHEMA #{SCHEMA} TO #{username}")
```
Comment (imtiazAhmd) [inline lib/analytics_creator.rb:38]: should we not just be "grant select"ing - least privilege principl. [USAGE](https://www.postgresql.org/docs/9.1/sql-grant.html) may grant more than we need or not enough to select from views?!

File: lib/analytics_creator.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    roles = execute("select * from pg_catalog.pg_roles where rolname='#{username}'")
+    return if roles.rows.any?
+
+    execute("CREATE ROLE #{username} WITH LOGIN PASSWORD '#{password}'")
+    execute("GRANT CONNECT ON DATABASE #{database_name} TO #{username}")
+    execute("CREATE SCHEMA IF NOT EXISTS #{schema}")
+    execute("GRANT USAGE ON SCHEMA #{schema} TO #{username}")
+    execute("grant select on all tables in schema #{schema} to #{username}")
+    execute("alter default privileges in schema #{schema} grant select on tables to #{username}")
+    execute("ALTER ROLE #{username} SET search_path TO #{schema}")
```
Comment (imtiazAhmd) [inline lib/analytics_creator.rb:54]: be good to understand why all of these are necessary

File: .env.development (hunk 2 above)
Comment (dwhenry) [reply .env.development:27]: yes this part is a WIP as I can't get it to work... :(

File: lib/analytics_creator.rb (hunk 3 above)
Comment (dwhenry) [reply lib/analytics_creator.rb:45]: Usage is need to access the schema... :) without it the account can't do anything based off:

https://stackoverflow.com/questions/17338621/what-does-grant-usage-on-schema-do-exactly

File: lib/analytics_creator.rb (hunk 4 above)
Comment (dwhenry) [reply lib/analytics_creator.rb:61]: it's not, but I'm trying to decide if it is worth leaving in so this would be used if scenic supported multiple DB (then you could:
* create an analytics schema
* give the analytics users full permission to view tables in the schema
* remove the rake job to set access rights

I will most likely remove it...

File: .env.development (hunk 2 above)
Comment (patrick-laa) [reply .env.development:27]: I think your issue is that `rds-postgresql-instance-output` already exists as a secret auto-created by cloud platform. You're not going to be able to re-create it. I think you'd have to create a new secret in AWS secret manager or add the new keys to the  existing `app-secrets` secret that's there.

File: db/schema.rb [hunk 5]
```diff
@@ -38,4 +40,27 @@
   end
 
   add_foreign_key "application_version", "application", name: "application_version_application_id_fkey"
+
+  create_view "all_events", sql_definition: <<-SQL
```
Comment (jogilder) [inline db/schema.rb:44]: Do we need these create statements and the new schema now that we have the migrations and other SQL files?

I thought everything was going to be part of the public schema.

File: app/jobs/notify_subscriber.rb [hunk 6]
```diff
@@ -35,7 +35,7 @@ def send_message_to_webhook(webhook_url, submission)
   end
 
   def delete_subscriber(subscriber)
-    deletion_threshold = ENV["SUBSCRIBER_FAILED_ATTEMPT_DELETION_THRESHOLD"]
+    deletion_threshold = ENV.fetch("SUBSCRIBER_FAILED_ATTEMPT_DELETION_THRESHOLD", nil)
```
Comment (jogilder) [inline app/jobs/notify_subscriber.rb:38]: Can we include details of the purpose of these changes in the pull request? They don't seem to be related to the ticket.

File: spec/rails_helper.rb [hunk 7]
```diff
@@ -21,7 +21,7 @@
 # directory. Alternatively, in the individual `*_spec.rb` files, manually
 # require only the support files necessary.
 #
-Dir[Rails.root.join("spec/support/**/*.rb")].sort.each { |f| require f }
+Dir[Rails.root.join("spec/support/**/*.rb")].each { |f| require f }
```
Comment (jogilder) [inline spec/rails_helper.rb:24]: What does this do?

File: app/jobs/notify_subscriber.rb (hunk 6 above)
Comment (patrick-laa) [reply app/jobs/notify_subscriber.rb:38]: This change doesn't actually do anything - in ruby, `[x]` is the same thing as `.fetch(x, nil)`.  I'm guessing this David has some code formatter plugin in his IDE that's configured to prefer fetch?

File: db/schema.rb (hunk 5 above)
Comment (patrick-laa) [reply db/schema.rb:44]: So in Rails the `schema.rb` contains a full snapshot of what the database structure should be. The migrations tell the story of how we got to the current snapshot, and how to get from a previous state to the current one. scenic adds a layer of extra redundancy because instead of altering views, it deletes and recreates them, and keeps a record of every version of every view. So it does mean every view definition in the schema file will be duplicated in db/views/myview_v[normally_latest].sql, but I think this redundancy is fine, as we've got Rails/scenic working to ensure that the right things are synced in the right ways.

File: spec/rails_helper.rb (hunk 7 above)
Comment (patrick-laa) [reply spec/rails_helper.rb:24]: This line ensures that test helper functions are correctly imported when running tests. The `sort` step was redundant as the order in which they're imported doesn't matter. I don't know what prompted removing it here.

**Review by patrick-laa — APPROVED** [review]

File: db/schema.rb [hunk 8]
```diff
@@ -10,7 +10,9 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.1].define(version: 2024_04_15_143554) do
+ActiveRecord::Schema[7.1].define(version: 2024_06_11_092435) do
+  create_schema "analytics"
```
Comment (jogilder) [inline db/schema.rb:14]: Why do we have an additional schema?

**Review by jogilder — APPROVED** [review]
Just the one comment on the schema but happy with the rest. Useful documentation as well.

File: db/schema.rb (hunk 8 above)
Comment (jsugarman) [reply db/schema.rb:14]: I believe this creeped in from something David may have tried previously. will take a look at removing or using
